onnxruntime/core/session/onnxruntime_c_api_ep.cc

```diff
@@ -0,0 +1,733 @@
+          auto added_graph_input = std::find_if(graph_inputs_including_initializers.begin(),
+                                                graph_inputs_including_initializers.end(),
+                                                [&name](const NodeArg* entry) { return entry->Name() == name; });
+
+          if (added_graph_input == graph_inputs_including_initializers.end()) {
+            if (context) {
+              auto type_proto = std::make_unique<ONNX_NAMESPACE::TypeProto>();
+              type_proto->CopyFrom(*(input->TypeAsProto()));
+              auto& n_input = top_level_graph->GetOrCreateNodeArg(name, type_proto.get());
+              context->manually_added_graph_inputs[n_input.Name()] = &n_input;
+//              LOGS_DEFAULT(VERBOSE) << "[TensorRT EP] \t" << n_input.Name() << " is added as an explicit input into the newly built graph";
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+static void SetAllGraphInputs(Graph& graph, std::unordered_map<std::string, std::unique_ptr<SubGraphContext2>>& subgraph_context_map) {
+  // If ORT TRT doesn't manully set graph input in TensorrtExecutionProvider::SetGraphOuterScopeValuesAndInputs(),
+  // Graph::Resolve() will help set graph inputs in Graph::SetGraphInputsOutputs(), so no need to set graph inputs here.
+  std::string unique_graph_name = GetUniqueGraphName(graph);
+  if (subgraph_context_map.find(unique_graph_name) == subgraph_context_map.end() ||
+      subgraph_context_map[unique_graph_name].get()->manually_added_graph_inputs.size() == 0) {
+    return;
+  }
+
+  SubGraphContext2* context = subgraph_context_map[unique_graph_name].get();
+  std::vector<const NodeArg*> graph_inputs_including_initializers;
+  std::unordered_set<std::string> graph_inputs_including_initializers_set;
+
+  for (const auto& entry : context->inputs_and_initializers) {
+    graph_inputs_including_initializers.push_back(entry.second);
+    graph_inputs_including_initializers_set.insert(entry.first);
+  }
+
+  for (const auto& entry : context->manually_added_graph_inputs) {
+    if (graph_inputs_including_initializers_set.find(entry.first) == graph_inputs_including_initializers_set.end()) {
+      graph_inputs_including_initializers.push_back(entry.second);
+      graph_inputs_including_initializers_set.insert(entry.first);
+    }
+  }
+
+  for (const auto& node_arg : graph.GetInputsIncludingInitializers()) {
+    if (graph_inputs_including_initializers_set.find(node_arg->Name()) == graph_inputs_including_initializers_set.end()) {
+      graph_inputs_including_initializers.push_back(node_arg);
+      graph_inputs_including_initializers_set.insert(node_arg->Name());
+    }
+  }
+
+  graph.SetInputs(graph_inputs_including_initializers);
+}
+
+ORT_API_STATUS_IMPL(OrtGraphApis::OrtGraph_GetSubGraph, const OrtGraphViewer* graph, const int node_num, const size_t* node_indices, _Outptr_ const OrtGraphViewer** subgraph) {
```

Comment:
I think it's debatable whether this api should return the subgraph's "GraphViewer" object.

The GraphViewer object holds a reference to the Graph object which is in memory. It means this api needs to keep Graph object alive after the function exits (so that's why it uses `new Model()` below), and it's hard to track the Graph object and release it, which leads to mem leak.

I'm thinking is it possible to return OrtGraph/Graph object instead?
In the case of TRT EP, once TRT EP has the OrtGraph object and it can:

- OrtGraph ----(new ep api?)----> OrtGraphViewer ----(OrtGraph_SerializeToArray())---->serialized string buffer
- OrtGraph ----(new ep api?)---->serialized string buffer

Reply:
It's a good point, we can just return OrtGraph, so that we don't need both OrtGraph_ReleaseGraph and OrtGraph_ReleaseGraphViewer. The problem is that we need to handle this subgraph so it requires more code change if it is in the form of OrtGraph. I added a TODO in the code for the future work